lib/ansible/modules/cloud/amazon/ec2_vpc_subnet_facts.py

```diff
@@ -74,76 +78,169 @@
+            returned: always
+            type: complex
+            contains:
+                association_id:
+                    description: The association ID
+                    returned: always
+                    type: string
+                ipv6_cidr_block:
+                    description: The IPv6 CIDR block that is associated with the subnet.
+                    returned: always
+                    type: string
+                ipv6_cidr_block_state:
+                    description: A hash/dict that contains a single item. The state of the cidr block association.
+                    returned: always
+                    type: dict
+                    contains:
+                        state:
+                            description: The CIDR block association state.
+                            returned: always
+                            type: string
 '''
 
+import traceback
+from ansible.module_utils.basic import AnsibleModule
+from ansible.module_utils.ec2 import (
+    boto3_conn,
+    ec2_argument_spec,
+    get_aws_connection_info,
+    AWSRetry,
+    HAS_BOTO3,
+    boto3_tag_list_to_ansible_dict,
+    camel_dict_to_snake_dict,
+    ansible_dict_to_boto3_filter_list
+)
+
 try:
-    import boto.vpc
-    from boto.exception import BotoServerError
-    HAS_BOTO = True
+    import botocore
 except ImportError:
-    HAS_BOTO = False
+    pass  # caught by imported HAS_BOTO3
 
-from ansible.module_utils.basic import AnsibleModule
-from ansible.module_utils.ec2 import AnsibleAWSError, connect_to_aws, ec2_argument_spec, get_aws_connection_info
 
+@AWSRetry.exponential_backoff()
+def describe_subnets_with_backoff(connection, subnet_ids, filters):
+    """
+    Describe Subnets with AWSRetry backoff throttling support.
 
-def get_subnet_info(subnet):
+    connection  : boto3 client connection object
+    subnet_ids  : list of subnet ids for which to gather facts
+    filters     : additional filters to apply to request
+    """
+    return connection.describe_subnets(SubnetIds=subnet_ids, Filters=filters)
 
-    subnet_info = {'id': subnet.id,
-                   'availability_zone': subnet.availability_zone,
-                   'available_ip_address_count': subnet.available_ip_address_count,
-                   'cidr_block': subnet.cidr_block,
-                   'default_for_az': subnet.defaultForAz,
-                   'map_public_ip_on_launch': subnet.mapPublicIpOnLaunch,
-                   'state': subnet.state,
-                   'tags': subnet.tags,
-                   'vpc_id': subnet.vpc_id}
 
-    return subnet_info
+def describe_subnets(connection, module):
+    """
+    Describe Subnets.
 
+    module  : AnsibleModule object
+    connection  : boto3 client connection object
+    """
+    # collect parameters
+    filters = ansible_dict_to_boto3_filter_list(module.params.get('filters'))
+    subnet_ids = module.params.get('subnet_ids')
 
-def list_ec2_vpc_subnets(connection, module):
+    if subnet_ids is None:
```

Comment:
This should never happen as `subnet_ids` is empty list by default, according to the argument_spec.

Reply:
If you specify the `subnet_ids` parameter with no value then it comes in as None. i.e. A task like this:
```
    - ec2_vpc_subnet_facts:
        subnet_ids:
        filters:
            vpc-id: vpc-12f23138
```
will cause this error:
```
botocore.exceptions.ParamValidationError: Parameter validation failed:
Invalid type for parameter SubnetIds, value: None, type: <type 'NoneType'>, valid types: <type 'list'>, <type 'tuple'>
```

It seems you only get the empty list default if the parameter is not specified at all.

Reply:
Not sure why you'd ever do that, but sure, that would happen :)